### PR TITLE
Update Ember, Ember Data, Ember QUnit types.

### PR DESCRIPTION
**Pull request by chriskrycho**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
    - <https://emberjs.com/api/ember/2.17/classes/Mixin>
    - <https://emberjs.com/api/ember/2.16/classes/Route/methods/afterModel?anchor=afterModel>
    - <https://emberjs.com/api/ember/2.16/classes/Route/methods/beforeModel?anchor=beforeModel>
    - <https://emberjs.com/api/ember/2.16/classes/@ember%2Fobject%2Fcomputed/methods/and?anchor=and>
    - <https://emberjs.com/api/ember/2.16/classes/@ember%2Fobject%2Fcomputed/methods/or?anchor=or>
    - <https://emberjs.com/api/ember-data/2.16/classes/DS.Store/methods/adapterFor?anchor=adapterFor>
    - <https://emberjs.com/api/ember-data/2.16/classes/DS.Store/methods/serializerFor?anchor=serializerFor>
    - <https://emberjs.com/api/ember-data/2.16/classes/DS.Model/methods/serialize?anchor=serialize>
    - <https://github.com/emberjs/ember-qunit/blob/master/README.md>

(Derived from work I did testing with my own large app; see [here](https://github.com/typed-ember/ember-typings/issues/69).)

### CODE REVIEW

**Comment by chriskrycho** [comment]
Will push a fix with the pre-function parens removed momentarily. Auto-formatting. 🙄

File: types/ember-data/index.d.ts [hunk 1]
```diff
@@ -1077,7 +1078,7 @@ declare module "ember-data" {
             /**
              * Returns the model class for the particular `modelName`.
              */
-            modelFor(modelName: string): Model;
+            modelFor<M extends Model>(modelName: string): M;
```
Comment (paulvanbrenk) [inline types/ember-data/index.d.ts]: This doesn't seem right. and the CI should have warned you here.

https://github.com/Microsoft/dtslint/blob/master/docs/no-unnecessary-generics.md

File: types/ember/index.d.ts [hunk 2]
```diff
@@ -107,7 +111,13 @@ declare module 'ember' {
 
     type ObserverMethod<Target, Sender> =
         | (keyof Target)
-        | ((this: Target, sender: Sender, key: keyof Sender, value: any, rev: number) => void);
+        | ((
```
Comment (paulvanbrenk) [inline types/ember/index.d.ts]: changes like these make it very hard to review this PR, it's hard to see what the actual changes are. I would prefer you revert all the extra linebreaks you add in this PR... and if you really care about them make a separate PR with just those changes.

File: types/ember-data/index.d.ts [hunk 3]
```diff
@@ -472,7 +473,7 @@ declare module "ember-data" {
             /**
              * Same as `deleteRecord`, but saves the record immediately.
              */
-            destroyRecord(options: {}): Promise<any>;
+            destroyRecord(options: {}): RSVP.Promise<any>;
```
Comment (paulvanbrenk) [inline types/ember-data/index.d.ts]: I couldn't find in the docs you linked why this change was necessary. (Or any other changes from `Promise` to `RSVP.Promise`)

**Review by paulvanbrenk — CHANGES_REQUESTED** [review]

File: types/ember-data/index.d.ts (hunk 1 above)
Comment (chriskrycho) [reply types/ember-data/index.d.ts]: For what it's worth, this one isn't actually an unnecessary generic: it's there to allow the caller specify *which* `Model` class gets returned. (This is an unfortunate workaround for the way the Ember Data typings work at present; we're working on an alternative approach, but for present it's what we have.)

Invocation looks like this:

```ts
const someInstance = modelFor<SomeModel>('some-model');
```

This is the only way, until we land that alternative approach, to get the correct types available in this and similar methods below.

File: types/ember-data/index.d.ts (hunk 3 above)
Comment (chriskrycho) [reply types/ember-data/index.d.ts]: All `Promise` instances *everywhere* in Ember (including Ember Data) are `RSVP.Promise` instances. The best reference I can come up with for that is [here](https://guides.emberjs.com/v2.16.0/contributing/repositories/#toc_code-rsvp-code):

> rsvp.js - Implementation of the of Promises/A+ spec used by Ember.

**Review by chriskrycho — COMMENTED** [review]
I've added comments on the non-whitespace issues; I'll fix the whitespace issues and push a change. (I thought I already *had* caught all of those, but missed them in a couple spots. Prettier is great until it isn't.)

**Comment by chriskrycho** [comment]
I've now reverted all the whitespace/comma/etc. Prettier-induced changes, @paulvanbrenk. Thanks!
